File: src/main/platform/macos/terminator.ts
/**
 * macOS application termination.
 *
 * Graceful quit uses `NSRunningApplication.terminate()`, which posts a proper
 * "quit" request (equivalent to ⌘Q) so apps can save state and close windows
 * cleanly — far better than signalling the process directly. Force quit is a
 * last resort, only used when the user explicitly opts in for unresponsive
 * apps. The Finder shell is never quit; its windows are closed via AppleScript.
 */
import type { RunningApp } from '@shared/types'
import type { ApplicationTerminator } from '../types'
import { run, runJxa, safeJsonForSource } from '../util'

function quitScript(pids: number[], force: boolean): string {
  return `
ObjC.import('AppKit');
(function () {
  const pids = ${safeJsonForSource(pids)};
  const results = [];
  for (const pid of pids) {
    const app = $.NSRunningApplication.runningApplicationWithProcessIdentifier(pid);
    if (!app || app.isNil()) { results.push([pid, 'gone']); continue; }
    if (app.isTerminated) { results.push([pid, 'gone']); continue; }
    const ok = ${force ? 'app.forceTerminate' : 'app.terminate'};
    results.push([pid, ok ? 'requested' : 'failed']);
  }
  return JSON.stringify(results);
})();
`
}

export class MacTerminator implements ApplicationTerminator {
  async requestQuit(app: RunningApp): Promise<void> {
    if (!app.pids.length) return
    try {
      await runJxa(quitScript(app.pids, false), { timeoutMs: 8000 })
    } catch {
      // Fall back to a polite SIGTERM if the bridge is unavailable.
      for (const pid of app.pids) {
        try {
          process.kill(pid, 'SIGTERM')
        } catch {
          /* already gone */
        }
      }
    }
  }

  async forceQuit(app: RunningApp): Promise<void> {
    if (!app.pids.length) return
    try {
      await runJxa(quitScript(app.pids, true), { timeoutMs: 8000 })
    } catch {
      for (const pid of app.pids) {
        try {
          process.kill(pid, 'SIGKILL')
        } catch {
          /* already gone */
        }
      }
    }
  }

  async closeFileManagerWindows(): Promise<void> {
    try {
      await run('osascript', ['-e', 'tell application "Finder" to close every window'], { timeoutMs: 6000 })
    } catch {
      // Automation permission may be denied; closing Finder windows is optional.
    }
  }
}

File: src-tauri/src/platform/macos/terminator.rs
//! macOS application termination.
//!
//! Graceful quit uses `NSRunningApplication.terminate()`, which posts a
//! proper "quit" request (equivalent to Cmd+Q) so apps can save state and
//! close windows cleanly — far better than signalling the process directly.
//! Force quit is a last resort, only used when the user explicitly opts in
//! for unresponsive apps. The Finder shell is never quit; its windows are
//! closed via AppleScript.

use super::{run_applescript, run_jxa};
use crate::platform::ApplicationTerminator;
use crate::types::RunningApp;

fn quit_script(pids: &[u32], force: bool) -> String {
    let pids_json = serde_json::to_string(pids).unwrap_or_else(|_| "[]".into());
    let method = if force { "app.forceTerminate" } else { "app.terminate" };
    format!(
        r#"
ObjC.import('AppKit');
(function () {{
  const pids = {pids_json};
  for (const pid of pids) {{
    const app = $.NSRunningApplication.runningApplicationWithProcessIdentifier(pid);
    if (!app || app.isNil() || app.isTerminated) continue;
    {method};
  }}
  return '';
}})();
"#
    )
}

pub struct MacTerminator;

impl ApplicationTerminator for MacTerminator {
    fn request_quit(&self, app: &RunningApp) {
        if app.pids.is_empty() {
            return;
        }
        if run_jxa(&quit_script(&app.pids, false)).is_err() {
            // Fall back to a polite SIGTERM if the bridge is unavailable.
            for &pid in &app.pids {
                unsafe {
                    libc::kill(pid as i32, libc::SIGTERM);
                }
            }
        }
    }

    fn force_quit(&self, app: &RunningApp) {
        if app.pids.is_empty() {
            return;
        }
        if run_jxa(&quit_script(&app.pids, true)).is_err() {
            for &pid in &app.pids {
                unsafe {
                    libc::kill(pid as i32, libc::SIGKILL);
                }
            }
        }
    }

    fn close_file_manager_windows(&self) {
        let _ = run_applescript(r#"tell application "Finder" to close every window"#);
    }
}
